src/lib/ContractInfo.ts

```diff
@@ -1,11 +1,182 @@
+
+export abstract class ContractInfo {
+    public static fromObject(o: ContractInfoEntry): VestingContractInfo | HashedTimeLockedContractInfo {
+        switch (o.type) {
+            case Nimiq.Account.Type.VESTING:
+                return VestingContractInfo.fromObject(o as VestingContractInfoEntry);
+            case Nimiq.Account.Type.HTLC:
+                return HashedTimeLockedContractInfo.fromObject(o as HashedTimeLockedContractInfoEntry);
+            default: throw new Error('Unknown contract type: ' + o.type);
+        }
+    }
+
+    public walletId?: string;
+
+    constructor(
+        public type: Nimiq.Account.Type,
+        public label: string,
+        public address: Nimiq.Address,
+        public balance?: number,
+    ) {}
+
+    public get userFriendlyAddress(): string {
+        return this.address.toUserFriendlyAddress();
+    }
+
+    public abstract toObject(): ContractInfoEntry;
+
+    public abstract toContractType(): Contract;
 }
 
-export interface ContractInfo {
-    address: Nimiq.Address;
+export class VestingContractInfo extends ContractInfo {
+    public static fromObject(o: VestingContractInfoEntry): VestingContractInfo {
+        return new VestingContractInfo(
+            o.label,
+            new Nimiq.Address(o.address),
+            new Nimiq.Address(o.owner),
+            o.start,
+            o.stepAmount,
+            o.stepBlocks,
+            o.totalAmount,
+            o.balance,
+        );
+    }
+
+    public constructor(
+        public label: string,
+        public address: Nimiq.Address,
+        public owner: Nimiq.Address,
+        public start: number,
+        public stepAmount: number,
+        public stepBlocks: number,
+        public totalAmount: number,
+        public balance?: number,
+    ) {
+        super(Nimiq.Account.Type.VESTING, label, address, balance);
+    }
+
+    public toObject(): VestingContractInfoEntry {
+        return {
+            type: this.type,
+            label: this.label,
+            address: new Uint8Array(this.address.serialize()),
+            owner: new Uint8Array(this.owner.serialize()),
+            start: this.start,
+            stepAmount: this.stepAmount,
+            stepBlocks: this.stepBlocks,
+            totalAmount: this.totalAmount,
+            balance: this.balance,
+        };
+    }
+
+    public toContractType(): VestingContract {
+        return {
+            type: Nimiq.Account.Type.VESTING,
+            label: this.label,
+            address: this.userFriendlyAddress,
+            owner: this.owner.toUserFriendlyAddress(),
+            start: this.start,
+            stepAmount: this.stepAmount,
+            stepBlocks: this.stepBlocks,
+            totalAmount: this.totalAmount,
+        };
+    }
+
+    public calculateAvailableAmount(height: number, currentBalance = this.totalAmount) {
+        return (currentBalance - this.totalAmount)
```

Comment:
I don't understand this calculation (above line especially), can you please comment?

Reply:
Does this explain it well: https://github.com/nimiq/accounts/pull/186/commits/156484922af87b08486238b55aa52f6bcafbd2d6 ?